sort_image_files.sh

```diff
@@ -0,0 +1,223 @@
+    
+    # Prepare Filename Change that includes Date. Use modify_date, change format
+    # & use in filename.
+    # Given Format:  2018-10-09T18:42:41+00:00
+    # Wanted Format: 2015-09-02_07-09_0060.jpg
+    # 1. Replace last 3 chars, e.g., :17	With: nothing
+    # 2. Replace ALL : 	With: -
+    # 3. Replace ALL spaces  With: _
+    date_for_filename_change="${modify_date::-9}" # Result: 2018-10-09T18:42
+    date_for_filename_change="${date_for_filename_change//T/-}" # Result: 2018-10-09_18:42
+    date_for_filename_change="${date_for_filename_change//:/_}" # Result: 2018-10-09_18_42
+  
+    # Build onto new_file_name. Concat to front - year & month.
+    # Use: ${string:position:length}   On: 2018-10-09T18:42:41+00:00
+    year="${modify_date:0:4}"
+    month="${modify_date:5:2}"
+  fi
+
+  just_path=$(dirname "${a_file_name}")
+  just_filename=$(basename "${a_file_name}") # For path to move files into subdirectories
+
+  # 4 Cases for Options to Add Day & Rename Files:
+  # Case 1: Add Day AND Rename files.
+  # Case 2: Add Day. No Rename files.
+  # Case 3: No add Day. Rename files.
+  # Case 4: No add Day. No Rename files.
+
+  day="${date_for_date_change:6:2}"
+  path_with_year_month_day="${just_path}/${year}/${month}/${day}"
+  path_with_year_month="${just_path}/${year}/${month}"
+
+  if [[ "${clean_day_subdir_also}" == 'y' ]] && [[ "${clean_rename_files_also}" == 'y' ]]; then
+    # Add Day subsubdirectories so Year-Month-Day AND Rename Files with IMG in them.
+    path_for_subdirs_creation="${path_with_year_month_day}"
+    renamed_file_and_path="${a_file_name/IMG/$date_for_filename_change}"
+    just_filename=$(basename "${renamed_file_and_path}") # For path to move files into subdirectories
+    new_dir_and_filename="${just_path}/${year}/${month}/${day}/${just_filename}"
+  elif [ "${clean_day_subdir_also}" == 'y' ]; then # ONLY add Day subdirectories.
+    path_for_subdirs_creation="${path_with_year_month_day}"
+    new_dir_and_filename="${just_path}/${year}/${month}/${day}/${just_filename}"
+  elif [ "${clean_rename_files_also}" == 'y' ]; then # ONLY rename files.
+    # Replace IMG in filename with value in $datestring_for_filename, which is
+    # in the format: YYYY-MM-DD_HH-MM, e.g., 2016-01-27_08-15.
+    new_dir_and_filename="${a_file_name/IMG/$date_for_filename_change}"
+    renamed_file_and_path="${a_file_name/IMG/$date_for_filename_change}"
+    echo "renamed_file_and_path: " "${renamed_file_and_path}"
+    path_for_subdirs_creation="${path_with_year_month}"
+    
+    # Get just filename from renamed_file_and_path.
+    just_filename=$(basename "${renamed_file_and_path}") # For path to move files into subdirectories
+    new_dir_and_filename="${just_path}/${year}/${month}/${just_filename}"
+  else # ONLY make year-month subdirectories.
+    path_for_subdirs_creation="${path_with_year_month}"
+    new_dir_and_filename="${just_path}/${year}/${month}/${just_filename}"
```

Comment:
And this is the same as

```shell
new_dir_and_filename="${path_for_subdirs_creation}/${just_filename}"
```

That means that in each of the four cases, you can assign the same expression to `new_dir_and_filename` -- and that means you can move the assignment

```shell
new_dir_and_filename="${path_for_subdirs_creation}/${just_filename}"
```
after your big `if` block. After you do that, I think you will notice that you can split your big four-case `if-else` block into two smaller `if-else` blocks, one to figure out what `path_for_subdirs_creation` is, and one to figure out what `just_filename` is.

Reply:
@kcharter Good suggestion. Eliminates several lines of code from cases. Thanks.

Reply:
@kimlew looks like you simplified things nicely.